src/ruis/render/scene/scene_renderer.cpp

```diff
@@ -0,0 +1,190 @@
+
+	prepare_fullscreen_quad_vao();
+}
+
+void scene_renderer::set_scene_scaling_factor(ruis::real scene_scaling_factor)
+{
+	this->scene_scaling_factor = scene_scaling_factor;
+}
+
+void scene_renderer::set_scene(std::shared_ptr<ruis::render::scene> scene_v)
+{
+	this->scene_v = scene_v;
+}
+
+void scene_renderer::set_environment_cube(std::shared_ptr<ruis::res::texture_cube> texture_environment_cube)
+{
+	this->texture_environment_cube = texture_environment_cube;
+}
+
+void scene_renderer::set_external_camera(std::shared_ptr<ruis::render::camera> cam)
+{
+	external_camera = cam;
+}
+
+void scene_renderer::render(const ruis::rect& dims, const ruis::mat4& viewport_matrix)
+{
+	if (!scene_v)
+		return;
+
+	auto cam = external_camera ? external_camera : scene_v.get()->active_camera;
+
+	if (!cam)
+		return;
+
+	float aspect = dims.d.x() / dims.d.y();
+	projection_matrix = viewport_matrix * cam->get_projection_matrix(aspect);
+
+	view_matrix = cam->get_view_matrix();
+
+	constexpr ruis::vec4 default_light_position{2, 4, -2, 1};
+	constexpr ruis::vec3 default_light_intensity{2, 2, 2};
+
+	if (scene_v.get()->lights.size() > 0) // scene has at least 1 light
+	{
+		main_light = scene_v.get()->lights[0].get();
+	} else {
+		main_light.pos = default_light_position;
+		main_light.intensity = default_light_intensity;
+	}
+
+	{
+		auto& r = this->context_v.get().renderer.get();
+		bool depth = r.is_depth_enabled();
+		r.enable_depth(false);
+		utki::scope_exit scope_exit([&r, depth]() {
+			r.enable_depth(depth);
+		});
+		render_environment();
+	}
+
+	ruis::mat4 root_model_matrix;
+	root_model_matrix.set_identity();
+	root_model_matrix.scale(scene_scaling_factor, scene_scaling_factor, scene_scaling_factor);
+
+	for (const auto& node_v : scene_v->nodes) {
+		this->render_node(node_v, root_model_matrix);
+	}
+}
+
+void scene_renderer::prepare_fullscreen_quad_vao()
+{
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+	std::array<ruis::vec2, 4> pos = {
+		{{-1, -1}, {-1, 1}, {1, -1}, {1, 1}}
+	};
+
+	auto pos_vbo = this->context_v.get().renderer.get().factory->create_vertex_buffer(utki::make_span(pos));
+
+	// NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+	std::array<uint16_t, 36> indices = {
+		{0, 2, 1, 1, 2, 3}
+	};
+
+	auto indices_vbo = this->context_v.get().renderer.get().factory->create_index_buffer(utki::make_span(indices));
+
+	this->fullscreen_quad_vao =
+		context_v.get()
+			.renderer.get()
+			.factory->create_vertex_array({pos_vbo}, indices_vbo, ruis::render::vertex_array::mode::triangles)
+			.to_shared_ptr();
+}
+
+void scene_renderer::render_environment()
+{
+	carcockpit::application::inst().shader_skybox_v.render(
+		*fullscreen_quad_vao.get(),
+		view_matrix,
+		projection_matrix,
+		texture_environment_cube ? texture_environment_cube->tex() : texture_default_environment_cube->tex()
+	);
+}
+
+void scene_renderer::render_node(utki::shared_ref<node> n, ruis::mat4 parent_model_matrix)
```

Comment:
I think, arguments to this functions should be:
```
render_node(const node& n, ruis::mat4 parent_model_matrix)
```
because you don't pass ownership of the node to this function, and the node cannot be null, so usual reference should work here

Reply:
changed